gcp/modules/istio/main.tf

```diff
@@ -1,3 +1,56 @@
 terraform {
   backend "gcs" {}
 }
+
+variable "project_id" {}
+variable "serviceaccount_key" {}
+variable "infra_region" {}
+variable "nonce" {}
+
+provider "google" {
+  project     = "${var.project_id}"
+  credentials = "${var.serviceaccount_key}"
+}
+
+# This resource waits till GCP allocates IP address for Istio ingress gateway
+resource "null_resource" "ingress_ip_wait" {
+  triggers = {
+    nonce = "${var.nonce}"
+  }
+
+  provisioner "local-exec" {
+    command = <<EOF
+      COUNT=1
+      MAX_RETRIES=60
+      SLEEP_SEC=5
+      INGRESS_READY=false
+
+      while [ "$INGRESS_READY" != 'true' ] && [ "$COUNT" -le "$MAX_RETRIES" ]; do
+        echo "Waiting for istio-ingressgateway ip ($COUNT/$MAX_RETRIES)"
+        kubectl -n istio-system get svc istio-ingressgateway -o json | jq -e '.status.loadBalancer.ingress[0].ip'
+        [ "$?" -eq 0 ] && INGRESS_READY=true
+        # Sleep only if we're not ready
+        [ "$INGRESS_READY" != 'true' ] && sleep "$SLEEP_SEC"
+        COUNT=$((COUNT+1))
+      done
+    EOF
+  }
+}
+
+data "kubernetes_service" "istio-ingressgateway" {
+  metadata {
+    name      = "istio-ingressgateway"
+    namespace = "istio-system"
+  }
+
+  depends_on = ["null_resource.ingress_ip_wait"]
+}
+
+resource "google_compute_address" "istio-ingress" {
+  project      = "${var.project_id}"
+  region       = "${var.infra_region}"
+  name         = "istio-ingress"
+  address_type = "EXTERNAL"
+  address      = "${data.kubernetes_service.istio-ingressgateway.load_balancer_ingress.0.ip}"
+  network_tier = "PREMIUM"
```

Comment:
Do we really want premium tier? Not sure if we'll be getting much from its benefits atm.

Reply:
Premium is what we use now. I have commented on this in the commit message (931f455):

> Note on "PREMIUM" network tier: Premium is Google's default, it's the
tier we currently use and apart from implication on performance, it's
the only tier that is available across all regions and the only tier
that can be used with global IPs.

So not using it would be a change from the current state, and amongst others would limit us to only certain regions (https://cloud.google.com/network-tiers/docs/overview#regions_supporting_standard_tier). Since we're not changing current state and since it's a complex topic with many implications (incl. pricing, different routing and other limitations), let's discuss this separately.